feat(rust): rust sdk docs (#4412)
temporalio/documentation: docs/develop/rust/activities/basics.mdx

        Err(ActivityError::cancelled())
    }
}
```

Activity parameters should be serializable and deserializable using serde. Use `#[derive(Serialize, Deserialize)]` on your data types:

```rust
use serde::{Serialize, Deserialize};
use temporalio_macros::activities;
use temporalio_sdk::activities::{ActivityContext, ActivityError};

#[derive(Serialize, Deserialize)]
pub struct GreetingInput {
    pub greeting: String,
    pub name: String,
}

pub struct GreetingActivities;

#[activities]
impl GreetingActivities {
    #[activity]
    pub async fn compose_greeting(
        _ctx: ActivityContext,
        input: GreetingInput,
    ) -> Result<String, ActivityError> {
        Ok(format!("{} {}!", input.greeting, input.name))
    }
}
```

### Define Activity return values {#activity-return-values}

All data returned from an Activity must be serializable.

Activity return values are subject to payload size limits in Temporal. The default payload size limit is 2MB, and there is a hard limit of 4MB for any gRPC message size in the Event History transaction. Keep in mind that all return values are recorded in a [Workflow Execution Event History](/workflow-execution/event#event-history).

> The return type of an Activity is `Result<T, ActivityError>`. The `T` type must implement `Serialize`. Use `ActivityError::Retryable` for errors that should be retried, and `ActivityError::NonRetryable` for permanent failures:

Reviewer: This changed in the latest release
```suggestion
The return type of an Activity is `Result<T, ActivityError>`. The `T` type must implement `Serialize`. Use `ApplicationFailure::new` for errors that should be retried, and `ApplicationFailure::non_retryable` for permanent failures:
```

```rust
#[derive(serde::Serialize, serde::Deserialize, Debug, Clone, PartialEq, Eq, Hash)]
pub struct ProcessedData {
    pub processed: String,
}

#[activities]
impl MyActivities {
    #[activity]
    pub async fn process_data(
        _ctx: ActivityContext,
        input: String,
    ) -> Result<ProcessedData, ActivityError> {
        // If an error should be retried
        if !validate_input(&input) {
            return Err(ActivityError::Retryable { 
                source: "Invalid input format".into(), 
                explicit_delay: Some(Duration::from_secs(5)) 

>             });

Reviewer: If we want to make this a more basic use case we can just omit the next retry delay and suggest users use the built in blanket conversion:

```
return Err(anyhow::anyhow!("Invalid input format").into());
```

If we want to keep the explicit delay this would be correct:
```suggestion
            return Err(ApplicationFailure::builder(anyhow::anyhow!("Invalid input format"))
                .next_retry_delay(Duration::from_secs(5))
                .build()
                .into());
```

        }

        // If an error should not be retried
        if input.len() > 1000000 {
            return Err(ActivityError::NonRetryable(
                "Input too large".into()

>             ));

Reviewer: ```suggestion
            return Err(ApplicationFailure::non_retryable(
                anyhow::anyhow!("Input too large")
            ).into());
```

title: Activity basics - Rust SDK
sidebar_label: Activity basics
description: This section explains how to implement Activities with the Rust SDK
toc_max_heading_level: 4
keywords:
  - Rust SDK
tags:
  - Rust SDK
  - Temporal SDKs
---

## Develop a basic Activity {#develop-activities}

**How to develop a basic Activity using the Temporal Rust SDK.**

One of the primary things that Workflows do is orchestrate the execution of Activities. An Activity is a normal function or method execution that's intended to execute a single, well-defined action (either short or long-running), such as querying a database, calling a third-party API, or transcoding a media file. An Activity can interact with the world outside the Temporal Platform or use a Temporal Client to interact with a Temporal Service. For the Workflow to be able to execute the Activity, we must define the [Activity Definition](/activity-definition).

In the Temporal Rust SDK, Activities are defined using the `#[activities]` macro on an impl block, with individual Activity methods marked by the `#[activity]` macro.

```rust
use temporalio_sdk::activities::{ActivityContext, ActivityError};
use temporalio_macros::activities;

> #[activities]
> impl GreetingActivities {
>     #[activity]
>     pub async fn greet(_ctx: ActivityContext, name: String) -> Result<String, ActivityError> {
>         Ok(format!("Hello, {}!", name))
>     }
> 
>     #[activity]
>     pub async fn send_notification(_ctx: ActivityContext, message: String) -> Result<(), ActivityError> {
>         println!("Sending notification: {}", message);
>         Ok(())
>     }
> }
> 
> pub struct GreetingActivities;

Reviewer: It is awkward to declare the struct after the implementation
```suggestion
pub struct GreetingActivities;

#[activities]
impl GreetingActivities {
    #[activity]
    pub async fn greet(_ctx: ActivityContext, name: String) -> Result<String, ActivityError> {
        Ok(format!("Hello, {}!", name))
    }

    #[activity]
    pub async fn send_notification(_ctx: ActivityContext, message: String) -> Result<(), ActivityError> {
        println!("Sending notification: {}", message);
        Ok(())
    }
}
```

Author: Thanks! Rust is a new language for me so I'm not sure about the conventions.

```

The `#[activities]` macro marks the impl block as containing Activity definitions. Each method decorated with `#[activity]` becomes an Activity that can be invoked from a Workflow.

### Activity method signature requirements

Each Activity method must:

- Be `async` (return a future)
- Take `ActivityContext` as the first parameter
- Return `Result<T, ActivityError>` where `T` is the return type
- Be `pub` (public)

> 

Reviewer 2: This doesn't mention activities that can take `Arc<Self>` and be registered using an instance. We should have an example of that somewhere as well.

Author: I've added this bit about `Arc<Self>` to this section. @chris-olszewski Could you help me out with an example for this?

Reviewer: Will look at getting a better example, but here's one of our tests that makes use of `Arc<Self>` https://github.com/temporalio/sdk-rust/blob/0f2f1b4cf879203658e56a9f8c8f5e42706f2bbe/crates/sdk-core/tests/integ_tests/workflow_tests/activities.rs#L1048-L1066

Author: Thanks! I've got that test in as an example for now.

The `ActivityContext` parameter provides access to Activity execution information and capabilities like heartbeating. If you don't need it, you can use `_ctx` as a parameter name.

### Define Activity parameters {#activity-parameters}

**How to define Activity parameters using the Temporal Rust SDK.**

> There is no explicit limit to the total number of parameters that an [Activity Definition](/activity-definition) may support. However, there is a limit to the total size of the data that ends up encoded into a gRPC message Payload.

Reviewer: For Rust we actually have a hard cap of 6 parameters for activities.

Author: Thanks for the clarification.

A single argument is limited to a maximum size of 2 MB. And the total size of a gRPC message, which includes all the arguments, is limited to a maximum of 4 MB.

Also, keep in mind that all Payload data is recorded in the [Workflow Execution Event History](/workflow-execution/event#event-history) and large Event Histories can affect Worker performance.

We recommend that you use a single struct as an argument that wraps all the application data passed to Activities. This way you can change what data is passed to the Activity without breaking the function signature.

> Activity parameters must be serializable and deserializable using serde. Use `#[derive(Serialize, Deserialize)]` on your data types:

Reviewer: This isn't a hard requirement, but implementing the serde traits are the easiest way to ensure they work. If they have a protobuf message they can leverage [`ProstSerializable`](https://docs.rs/temporalio-common/latest/temporalio_common/data_converters/struct.ProstSerializable.html) for encoding without implementing the serde traits.

I am actively working on improving the ergonomics of using protobuf instead of serde, so I don't have a good sample yet.

Author: I'll soften the language here to say _should_ instead of _must_. Is this example ok for now? I'm not very familiar with serde or Rust in general. 😅

```rust
use serde::{Serialize, Deserialize};

#[derive(Serialize, Deserialize)]
pub struct GreetingInput {
    pub greeting: String,
    pub name: String,
}

#[activities]
impl GreetingActivities {
    #[activity]
    pub async fn compose_greeting(
        _ctx: ActivityContext,
        input: GreetingInput,
    ) -> Result<String, ActivityError> {
        Ok(format!("{} {}!", input.greeting, input.name))
    }
}
```

### Define Activity return values {#activity-return-values}

> **How to define Activity return values using the Temporal Rust SDK.**

Reviewer 2: I'd probably just combine this and the above section since they're really dealing with the same concept.

**How to customize your Activity Type using the Temporal Rust SDK.**

Activities have a Type that refers to the Activity name. The Activity name is used to identify Activity Types in the Workflow Execution Event History, Visibility Queries, and Metrics.

By default, the Activity name is the method name. You can customize it by providing a `name` parameter to the `#[activity]` macro:

```rust
#[activities]
impl GreetingActivities {
    #[activity(name = "compose_greeting")]
    pub async fn greet(_ctx: ActivityContext, name: String) -> Result<String, ActivityError> {
        Ok(format!("Hello, {}!", name))
    }

    #[activity(name = "send_email")]
    pub async fn send_notification(_ctx: ActivityContext, email: String) -> Result<(), ActivityError> {
        println!("Email to: {}", email);
        Ok(())
    }
}
```

## Activity Error Handling {#error-handling}

Activities return `Result<T, ActivityError>` with different error types for different failure scenarios:

- **`ActivityError::Retryable`** - Transient failure that should be retried according to the Activity's retry policy
- **`ActivityError::NonRetryable`** - Permanent failure that will not be retried

> - **`ActivityError::Cancelled`** - Activity was cancelled by the Workflow

Reviewer 2: This section should probably refer to what the default behavior is for errors bubbled up via `?`

(Which I believe is to convert them into a Retryable activity error)
